kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbPartitioningMetadataStore.java

```diff
@@ -0,0 +1,249 @@
+  public CompletionStage<T> findAsync(String path) {
+    return getOrCreateMetadataStore(findPartition(path)).getAsync(path);
+  }
+
+  @Deprecated
+  public T findSync(String path) {
+    try {
+      return findAsync(path).toCompletableFuture().get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new InternalMetadataStoreException("Error fetching node at path " + path, e);
+    }
+  }
+
+  public CompletionStage<Stat> updateAsync(T metadataNode) {
+    return getOrCreateMetadataStore(metadataNode.getPartition()).updateAsync(metadataNode);
+  }
+
+  public void updateSync(T metadataNode) {
+    try {
+      updateAsync(metadataNode)
+          .toCompletableFuture()
+          .get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new InternalMetadataStoreException("Error updating node: " + metadataNode, e);
+    }
+  }
+
+  public CompletionStage<Void> deleteAsync(T metadataNode) {
+    return getOrCreateMetadataStore(metadataNode.getPartition()).deleteAsync(metadataNode);
+  }
+
+  public void deleteSync(T metadataNode) {
+    try {
+      deleteAsync(metadataNode)
+          .toCompletableFuture()
+          .get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      throw new InternalMetadataStoreException(
+          "Error deleting node under at path: " + metadataNode.name, e);
+    }
+  }
+
+  @Deprecated
+  public CompletionStage<List<T>> getCachedAsync() {
+    return listAsync();
+  }
+
+  @Deprecated
+  public List<T> getCachedSync() {
+    return listSync();
+  }
+
+  public CompletableFuture<List<T>> listAsync() {
+    List<CompletableFuture<List<T>>> completionStages = new ArrayList<>();
+    for (Map.Entry<String, KaldbMetadataStore<T>> metadataStoreEntry :
+        metadataStoreMap.entrySet()) {
+      completionStages.add(metadataStoreEntry.getValue().getCachedAsync().toCompletableFuture());
+    }
+
+    return CompletableFuture.allOf(completionStages.toArray(new CompletableFuture[0]))
+        .thenApply(
+            (unused) ->
+                completionStages.stream()
+                    .map(f -> f.toCompletableFuture().join())
+                    .flatMap(List::stream)
+                    .collect(Collectors.toList()));
+  }
+
+  public List<T> listSync() {
+    try {
+      return listAsync().toCompletableFuture().get(DEFAULT_ZK_TIMEOUT_SECS, TimeUnit.SECONDS);
+    } catch (ExecutionException | InterruptedException | TimeoutException e) {
+      throw new InternalMetadataStoreException("Error listing nodes", e);
+    }
+  }
+
+  private KaldbMetadataStore<T> getOrCreateMetadataStore(String partition) {
+    return metadataStoreMap.computeIfAbsent(
+        partition,
+        (p1) -> {
+          LOG.info("Creating new metadata store for partition - {}", partition);
```

Comment:
will having the `path` to the output here be useful?

Reply:
Good idea! Added.